projects/ccd-case-ui-toolkit/src/lib/shared/components/palette/case-file-view/components/case-file-view-folder/case-file-view-folder.component.ts

```diff
@@ -0,0 +1,251 @@
+import { NestedTreeControl } from '@angular/cdk/tree';
+import { Component, Input, OnDestroy, OnInit } from '@angular/core';
+import { plainToClass } from 'class-transformer';
+import { Observable, of, Subscription } from 'rxjs';
+import {
+  CaseFileViewCategory,
+  CaseFileViewDocument,
+  CategoriesAndDocuments,
+  DocumentTreeNode
+} from '../../../../../domain/case-file-view';
+
+@Component({
+  selector: 'ccd-case-file-view-folder',
+  styleUrls: ['./case-file-view-folder.component.scss'],
+  templateUrl: './case-file-view-folder.component.html'
+})
+export class CaseFileViewFolderComponent implements OnInit, OnDestroy {
+  private static readonly UNCATEGORISED_DOCUMENTS_TITLE = 'Uncategorised documents';
+
+  @Input() public categoriesAndDocuments$: Observable<CategoriesAndDocuments>;
+
+  public nestedTreeControl: NestedTreeControl<DocumentTreeNode>;
+  public nestedDataSource: DocumentTreeNode[];
+  public categories: CaseFileViewCategory[] = [];
+  public categoriesAndDocumentsSubscription: Subscription;
+
+  private getChildren = (node: DocumentTreeNode) => of(node.children);
+  public hasNestedChild = (_: number, nodeData: DocumentTreeNode) => nodeData.children;
+
+  constructor() {
+    this.nestedTreeControl = new NestedTreeControl<DocumentTreeNode>(this.getChildren);
+  }
+
+  public ngOnInit(): void {
+    this.categoriesAndDocumentsSubscription = this.categoriesAndDocuments$
+      .subscribe(categoriesAndDocuments => {
+      // Using the mock data for now as we have to display the documents as well for demo purpose
+      const categories = this.loadCategories(); // categoriesAndDocuments.categories;
+      // Generate document tree data from categories
+      const treeData = this.generateTreeData(categories);
+      // Append uncategorised documents
+      if (categoriesAndDocuments.uncategorised_documents && categoriesAndDocuments.uncategorised_documents.length > 0) {
+        const uncategorisedDocuments = this.getUncategorisedDocuments(categoriesAndDocuments.uncategorised_documents);
+        treeData.push(uncategorisedDocuments);
+      }
+      // Initialise cdk tree with generated data
+      this.nestedDataSource = treeData;
+    });
+  }
+
+  public generateTreeData(categories: CaseFileViewCategory[]): DocumentTreeNode[] {
+    return categories.reduce((tree, node) => [
+      ...tree,
+      ...[
+        plainToClass(DocumentTreeNode, {
```

Comment:
try not to use the plainToClass function, as it's highly inefficient, especially if you know what class it should be beforehand.

Reply:
What do you think I should use instead -- new operator?
I preferred in this case as it's convenient to use -- i.e. it sorts creating instances for the children.